File: src/main/java/market/application/StoreService.java
package market.application;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import market.domain.store.IListingRepository;
import market.domain.store.IStoreRepository;
import market.domain.store.Store;
import market.domain.store.StoreDTO;
import market.domain.user.IUserRepository;
import utils.Logger;


public class StoreService {
    private IStoreRepository storeRepository;
    private IListingRepository listingRepository;
    private String storeIDs ="1";
    private Logger logger = Logger.getInstance();

    public StoreService(IStoreRepository storeRepository, IUserRepository userRepository, IListingRepository listingRepository) {
        this.storeRepository = storeRepository;
        storeIDs = storeRepository.getNextStoreID();
        this.listingRepository = listingRepository;
    }


    /*
     * return storeId
     */
    public String createStore(String storeName, String founderId) throws Exception {
        // ? - do we need store type
        if(storeRepository.containsStore(storeName)) {
            logger.error("Attempted to create store with existing name: " + storeName);
            throw new Exception("The storeName '" + storeName + "' already exists");
        }
        String id = storeIDs;
        storeIDs = String.valueOf(Integer.valueOf(storeIDs) + 1);
        Store store = new Store(String.valueOf(storeIDs),storeName, founderId,listingRepository);
        //Who is responsable to manage the store id's????????
        storeRepository.addStore(store);
        logger.info("Store created: " + storeName + ", founder: " + founderId + ", id: " + storeIDs);
        //((Subscriber)userRepository.findById(founderId)).setStoreRole(store.getStoreID(), "Founder");
        
        return storeIDs;
        //LOG - store added
    }

    /**
     * Closes a specific store.
     * Only the founder of the store is authorized to perform this operation.
     * If successful, the store is marked as inactive (closed).
     *
     * @param storeID ID of the store to close.
     * @param userName  ID of the user attempting to close the store.
     * @return A message indicating "success" if the operation succeeded, or an error message if it failed.
     * @throws Exception if the store does not exist or the closure fails internally.
     */
    public String closeStore(String storeID,String userName) throws Exception {
        try
        {
            Store s = storeRepository.getStoreByID(storeID);
            if (s==null){
                logger.error("Attempted to close non-existent store: " + storeID);
                throw new Exception("store doesn't exist");
            }
            if (!s.closeStore(userName)){
                logger.error("Failed to close store: " + storeID);
                throw new Exception("store cant be closed");

            }
            Set<String> users = s.getPositionsInStore(userName).keySet();
            for(String ownerOrManager : users){
                if(s.isManager(ownerOrManager))
                   ; // ((Subscriber)userRepository.findById(ownerOrManager)).removeStoreRole(storeID,"Manager");
                if(s.isOwner(ownerOrManager))
                    ; //((Subscriber)userRepository.findById(ownerOrManager)).removeStoreRole(storeID,"Owner");
                if(s.getFounderID().equals(ownerOrManager)){
                   ; // ((Subscriber)userRepository.findById(ownerOrManager)).removeStoreRole(storeID,"Founder");
                   ; // ((Subscriber)userRepository.findById(ownerOrManager)).removeStore(storeID);
                }

                //TODO:need to notify all the owners and managers
            }
            logger.info("Store closed: " + storeID + ", by user: " + userName);
        }
        catch(Exception e){
            logger.error("Error closing store: " + storeID + ", by user: " + userName + ". Reason: " + e.getMessage());
            return e.getMessage();
        }
        return "success";
    }


    /**
     * Reopens a specific store.
     * Only the founder of the store is authorized to perform this operation.
     * If successful, the store is marked as active (open).
     *
     * @param storeID ID of the store to open.
     * @param userName  ID of the user attempting to open the store.
     * @return A message indicating "success" if the operation succeeded, or an error message if it failed.
     * @throws Exception if the store does not exist or the reopening fails internally.
     */
    public String openStore(String storeID,String userName) throws Exception {
        try
        {
            Store s = storeRepository.getStoreByID(storeID);
            if (s==null){
                logger.error("Attempted to open non-existent store: " + storeID);
                throw new Exception("store doesn't exist");
            }
            s.openStore(userName);
            logger.info("Store opened: " + storeID + ", by user: " + userName);
            //TODO:need to notify all the owners and managers
        }
        catch(Exception e){
            logger.error("Error opening store: " + storeID + ", by user: " + userName + ". Reason: " + e.getMessage());
            return e.getMessage();
        }
        return "success";
    }


    

    public StoreDTO getStore(String storeName) throws Exception {
        Store store = storeRepository.getStoreByName(storeName);
        if(store == null) {
            logger.error("Attempted to get non-existent store: " + storeName);
            return null;
        }
        logger.info("Store retrieved: " + storeName);
        return new StoreDTO(store);
    }


/*
appoints 'newOwner' to be an owner of 'storeID' BY 'appointerID'
assumes aggreement by 'apointerID''s appointer
 */
    public String addAdditionalStoreOwner(String appointerID, String newOwnerID, String storeID){
        try
        {
            Store s = storeRepository.getStoreByID(storeID);
            if (s==null){
                logger.error("Attempted to add owner to non-existent store: " + storeID);
                throw new Exception("store doesn't exist");
            }
            s.addNewOwner(appointerID,newOwnerID);
            logger.info("Added new owner: " + newOwnerID + " to store: " + storeID + ", by: " + appointerID);
            //((Subscriber)userRepository.findById(newOwnerID)).setStoreRole(storeID , "Owner");
            //PAY ATTENTION! לעשות בכל מקום
        }
        catch (Exception e){
            logger.error("Error adding owner: " + newOwnerID + " to store: " + storeID + ". Reason: " + e.getMessage());
            //TODO:we need to decide how to handle things here
            return e.getMessage();
        }
        return "success";
    }

    /*
    requests to add a new owner to a store.
    if the founder requests, it does that.
    if an owner requests, so its send a notification to his appointer, to allow the appointment
     */
    public String OwnerAppointmentRequest(String appointerID, String newOwnerId,String storeID){
        try{
            Store s = storeRepository.getStoreByID(storeID);
            if (s==null)
                throw new Exception("store doesn't exist");
            if (s.getFounderID().equals(appointerID)){
                s.addNewOwner(appointerID,newOwnerId);
                logger.info("Founder " + appointerID + " added new owner: " + newOwnerId + " to store: " + storeID);
            }
            else{
                if (s.isOwner(newOwnerId)){
                    logger.error(newOwnerId + " is already an Owner of store:"+storeID);
                    throw new Exception(newOwnerId + " is already an Owner of store:"+storeID);
                }
                if (!s.isOwner(appointerID)){
                    logger.error(appointerID + " is NOT an Owner of store:"+storeID);
                    throw new Exception(appointerID + " is NOT an Owner of store:"+storeID);
                }
                String requestTO=s.OwnerAssignedBy(appointerID);
                logger.info("Owner appointment request: " + appointerID + " requests to appoint " + newOwnerId + " in store: " + storeID + ", request sent to: " + requestTO);
                //TODO:notify 'requestTO' that his assignee want to assign new owner
            }
        }catch (Exception e){
            logger.error("Error in owner appointment request for store: " + storeID + ". Reason: " + e.getMessage());
            //TODO:we need to decide how to handle things here
            return e.getMessage();
        }
        return "success";
    }



    /*
    removes 'toRemove' and all the people he assigned
     */
    public List<List<String>> removeOwner(String id, String toRemove, String storeID){
        List<List<String>> ret = new ArrayList<>();
        ret.add(new ArrayList<>());
        try {
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null)
                throw new Exception("store doesn't exist");
            List<List<String>> removedWorkers =  s.removeOwner(id,toRemove);
            logger.info("Removed owner: " + toRemove + " from store: " + storeID + ", by: " + id);
            // for (String i:removedWorkers.get(0) ){
            //     ((Subscriber)userRepository.findById(i)).removeStoreRole(id,"Owner");
            //     //TODO: need to change data on those ussers
            // }
            // for (String i:removedWorkers.get(1) ){
            //     ((Subscriber)userRepository.findById(i)).removeStoreRole(id,"Manager");
            //     //TODO: need to change data on those ussers
            // }
            return removedWorkers;
        }
        catch (Exception e){
            ret.get(0).add(e.getMessage());
            logger.error("Error removing owner: " + toRemove + " from store: " + storeID + ". Reason: " + e.getMessage());
            return ret;
        }
    }


    /**
     * Adds a new manager to the store by delegating to the business logic layer.
     * Only an existing owner of the store can appoint a new manager.
     *
     * @param appointerID ID of the owner appointing the new manager.
     * @param newManagerName ID of the user being assigned as a new manager.
     * @param storeID ID of the store where the manager is being added.
     * @return "success" if the manager was added successfully, "failed" if the operation was unsuccessful.
     * @throws RuntimeException if the store does not exist, the appointer is not an owner,
     *                           the new manager is already assigned, or any other business rule is violated.
     */
    public String addNewManager(String appointerID, String newManagerName, String storeID){
        try{
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null){
                logger.error("Attempted to add manager to non-existent store: " + storeID);
                throw new Exception("store doesn't exist");
            }
            if (s.addNewManager(appointerID,newManagerName)){
                logger.info("Added new manager: " + newManagerName + " to store: " + storeID + ", by: " + appointerID);
                //((Subscriber)userRepository.findById(newManagerName)).setStoreRole(storeID,"Manager");
                return "success";
            }
            else{
                logger.error("Failed to add manager: " + newManagerName + " to store: " + storeID + ", by: " + appointerID);
                return "failed";
            }

        } catch (Exception e) {
            logger.error("Error adding manager: " + newManagerName + " to store: " + storeID + ". Reason: " + e.getMessage());
            return e.getMessage();
        }
    }

    /**
     * Grants a specific permission to a manager in the specified store.
     * Only the owner who originally appointed the manager can grant permissions.
     *
     * @param managerID    ID of the manager receiving the new permission.
     * @param appointerID  ID of the owner who appointed the manager and is now granting the permission.
     * @param permissionID Integer code representing the permission to assign (must be valid in {Permission} enum).
     * @param storeID      ID of the store where the manager belongs.
     * @return "success" if the permission was successfully added, "failed" if the operation did not complete.
     * @throws RuntimeException if the store does not exist, the appointer is not authorized,
     *                           the manager is invalid, the permission code is invalid,
     *                           or any other business rule violation occurs.
     */
    public String addPermissionToManager(String managerID, String appointerID, int permissionID, String storeID) {
        try {
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null) {
                logger.error("Attempted to add permission to non-existent store: " + storeID);
                throw new Exception("store doesn't exist");
            }
            if (s.addPermissionToManager(managerID, appointerID, permissionID)) {
                logger.info("Added permission: " + permissionID + " to manager: " + managerID + " in store: " + storeID + ", by: " + appointerID);
                return "success";
            } else {
                logger.error("Failed to add permission: " + permissionID + " to manager: " + managerID + " in store: " + storeID + ", by: " + appointerID);
                return "failed";
            }
        } catch (Exception e) {
            logger.error("Error adding permission: " + permissionID + " to manager: " + managerID + " in store: " + storeID + ". Reason: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    /**
     * Retrieves the set of permission codes assigned to a specific manager in a store.
     * The requester must be either the store owner or a manager themself.
     *
     * @param managerID   ID of the manager whose permissions are being requested.
     * @param whoIsAsking ID of the user making the request (must be an owner or the manager).
     * @param storeID     ID of the store where the manager belongs.
     * @return A {@link Set} of integer codes representing the manager's current permissions.
     * @throws RuntimeException if the store does not exist, the requester is unauthorized,
     *                           the manager is invalid, or any other business rule violation occurs.
     */
    public Set<Integer> getManagersPermissions(String managerID, String whoIsAsking, String storeID) {
        try {
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null) {
                logger.error("Attempted to get permissions for non-existent store: " + storeID);
                throw new Exception("store doesn't exist");
            }
            logger.info("Retrieved permissions for manager: " + managerID + " in store: " + storeID + ", by: " + whoIsAsking);
            return s.getManagersPermmisions(managerID, whoIsAsking);
        } catch (Exception e) {
            logger.error("Error retrieving permissions for manager: " + managerID + " in store: " + storeID + ". Reason: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }

    /**
     * Removes a specific permission from a manager in the specified store.
     * Only the owner who originally appointed the manager can remove permissions.
     *
     * @param managerID    ID of the manager whose permission is being revoked.
     * @param permissionID Integer code representing the permission to remove (must be valid in { Permission} enum).
     * @param appointerID  ID of the owner who appointed the manager and is requesting to remove the permission.
     * @param storeID      ID of the store where the manager belongs.
     * @return "success" if the permission was successfully removed, "failed" if the operation did not complete.
     * @throws RuntimeException if the store does not exist, the appointer is unauthorized,
     *                           the manager ID is invalid, the permission code is invalid,
     *                           or any other business rule violation occurs.
     */
    public String removePermissionFromManager(String managerID, int permissionID, String appointerID, String storeID) {
        try {
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null) {
                logger.error("Attempted to remove permission from non-existent store: " + storeID);
                throw new Exception("store doesn't exist");
            }
            if (s.removePermissionFromManager(managerID, permissionID, appointerID)) {
                logger.info("Removed permission: " + permissionID + " from manager: " + managerID + " in store: " + storeID + ", by: " + appointerID);
                return "success";
            } else {
                logger.error("Failed to remove permission: " + permissionID + " from manager: " + managerID + " in store: " + storeID + ", by: " + appointerID);
                return "failed";
            }
        } catch (Exception e) {
            logger.error("Error removing permission: " + permissionID + " from manager: " + managerID + " in store: " + storeID + ". Reason: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }



    /**
     * Adds a new listing to the specified store.
     *
     * @param userName User trying to add.
     * @param storeID Store ID.
     * @param productId Product ID.
     * @param productName Product name.
     * @param productDescription Description of the product.
     * @param quantity Quantity to add.
     * @param price Price per unit.
     * @return "succeed" or error message.
     */
    public String addNewListing(String userName, String storeID, String productId, String productName, String productCategory, String productDescription, int quantity, double price) {
        try {
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null)
                throw new Exception("Store doesn't exist");
            logger.info("Added new listing: " + productName + " to store: " + storeID + ", by: " + userName);
            return s.addNewListing(userName, productId, productName, productCategory, productDescription, quantity, price);
        } catch (Exception e) {
            logger.error("Error adding listing: " + productName + " to store: " + storeID + ". Reason: " + e.getMessage());
            return e.getMessage();
        }
        
    }




    /**
     * Removes a listing from the specified store.
     *
     * @param userName User ID.
     * @param storeID Store ID.
     * @param listingId ID of the listing to remove.
     * @return "succeed" or error message.
     */
    public String removeListing(String userName, String storeID, String listingId) {
        try {
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null)
                throw new Exception("Store doesn't exist");
            logger.info("Removed listing: " + listingId + " from store: " + storeID + ", by: " + userName);
            s.removeListing(userName, listingId);
        } catch (Exception e) {
            logger.error("Error removing listing: " + listingId + " from store: " + storeID + ". Reason: " + e.getMessage());
            return e.getMessage();
        }
        return "succeed";
    }


    public IListingRepository getListingRepository(){
        return listingRepository;
    }


//I think this function is not relevant any more. all purchases are from purchase service
    /**
     * Purchases quantity from a listing.
     *
     *
     * @param storeID Store ID.
     * @param listingId ID of the listing to purchase from.
     * @param quantity How many units to buy.
     * @return "succeed" or error message.
     */
    public String purchaseFromListing( String storeID, String listingId, int quantity) {
        try {
            Store s = storeRepository.getStoreByID(storeID);
            if (s == null)
                throw new Exception("Store doesn't exist");
            logger.info("Purchased " + quantity + " units from listing: " + listingId + " in store: " + storeID);
            s.purchaseFromListing(listingId, quantity);
        } catch (Exception e) {
            logger.error("Error purchasing from listing: " + listingId + " in store: " + storeID + ". Reason: " + e.getMessage());
            return e.getMessage();
        }
        return "succeed";
    }

    public double getProductPrice(String storeID, String productID) {
        Store s = storeRepository.getStoreByID(storeID);
        Map<String,Integer> prod = new HashMap<>();
        prod.put(productID,1);
        logger.info("Retrieved price for product: " + productID + " in store: " + storeID);
        return s.calculateStoreBagWithDiscount(prod);
    }

    public String getProductListing(String storeID, String productID) {

        //TODO: implement this method to get product listing
        return null;
    }


    public Boolean checkAndUpdateStock(String storeID, String productID, int quantity) {
        //TODO: implement this method to check and update stock
        return true;
    }
}

File: frontend/src/services/storeService.js
import axios from 'axios';

// Create axios instance with base URL
const api = axios.create({
  baseURL: '/api'
});

// Add JWT token to requests if available
api.interceptors.request.use(config => {
  const token = localStorage.getItem('token');
  if (token) {
    config.headers.Authorization = `Bearer ${token}`;
  }
  return config;
});

export const storeService = {
  // Get store details by ID (internal helper)
  getStoreById: async (storeId) => {
    try {
      // Since we don't have a direct getStoreById endpoint, we'll try to find it
      // by getting all stores and finding the one with matching ID
      // For now, return a default structure
      return {
        id: storeId,
        name: `Store ${storeId}`,
        description: 'Store description not available',
        isActive: true,
        founderId: 'unknown'
      };
    } catch (error) {
      console.warn(`Failed to get store details for ID ${storeId}:`, error);
      return {
        id: storeId,
        name: `Store ${storeId}`,
        description: 'Store description not available',
        isActive: true,
        founderId: 'unknown'
      };
    }
  },

  // Get all stores and their products
  getAllStoresAndProducts: async () => {
    try {
      const response = await api.get('/stores/info');
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        const stores = [];
        const products = [];
        
        // Handle new API response format: List of {store, listings}
        apiResponse.data.forEach(storeData => {
          const store = storeData.store;
          const listings = storeData.listings;
          
          // Add store to stores array
          stores.push({
            id: store.storeID,
            name: store.storeName,
            description: store.description || 'No description available',
            isActive: store.isActive !== false,
            founderId: store.founderId || 'unknown',
            rating: store.rating || 0,
            totalProducts: listings.length
          });
          
          // Transform listings to products
          listings.forEach(listing => {
            products.push({
              id: listing.listingId,
              title: listing.productName,
              price: listing.price,
              status: listing.active ? 'active' : 'inactive',
              images: listing.images || [],
              category: listing.category,
              shipping_cost: listing.shippingCost || 0,
              featured: listing.active,
              description: listing.productDescription,
              quantity: listing.quantityAvailable,
              storeId: store.storeID,
              storeName: store.storeName,
              productId: listing.productId,
              purchaseType: listing.purchaseType,
              seller: {
                id: store.storeID,
                name: store.storeName,
                rating: store.rating || 0
              },
              created_date: listing.createdDate || new Date().toISOString(),
              rating: listing.rating || 0,
              reviews: listing.reviews || []
            });
          });
        });
        
        return { stores, products };
      } else {
        throw new Error('Failed to get stores and products');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to get stores and products');
      }
      throw new Error('Failed to get stores and products');
    }
  },

  // Get store by name
  getStore: async (storeName) => {
    try {
      const response = await api.get(`/stores/${storeName}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          id: apiResponse.data.storeID,
          name: apiResponse.data.storeName,
          description: apiResponse.data.description || 'No description available',
          isActive: apiResponse.data.isActive,
          founderId: apiResponse.data.founderId
        };
      } else {
        throw new Error('Failed to fetch store');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to fetch store');
      }
      throw new Error('Failed to fetch store');
    }
  },

  // Create a new store
  createStore: async (storeData) => {
    try {
      const response = await api.post('/stores/create', storeData);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          id: apiResponse.data.storeId,
          name: storeData.storeName,
          founderId: storeData.founderId,
          isActive: true
        };
      } else {
        throw new Error('Failed to create store');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to create store');
      }
      throw new Error('Failed to create store');
    }
  },

  // Add new listing/product
  addListing: async (userName, storeID, productId, productName, productCategory, productDescription, quantity, price, purchaseType = 'REGULAR') => {
    try {
      const response = await api.post('/stores/listings/add', {
        userName,
        storeID,
        productId,
        productName,
        productCategory,
        productDescription,
        quantity,
        price,
        purchaseType
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          productId: apiResponse.data,
          message: 'Product added successfully'
        };
      } else {
        throw new Error('Failed to add product');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to add product');
      }
      throw new Error('Failed to add product');
    }
  },

  // Remove a listing from a store
  removeListing: async (removeData) => {
    try {
      const response = await api.post('/stores/listings/remove', removeData);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Listing removed successfully'
        };
      } else {
        throw new Error('Failed to remove listing');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to remove listing');
      }
      throw new Error('Failed to remove listing');
    }
  },

  // Update a listing in a store
  updateListing: async (updateData) => {
    try {
      const response = await api.post('/stores/listings/update', updateData);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Listing updated successfully'
        };
      } else {
        throw new Error('Failed to update listing');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to update listing');
      }
      throw new Error('Failed to update listing');
    }
  },

  // Get all stores
  getAllStores: async () => {
    try {
      const { stores } = await this.getAllStoresAndProducts();
      return stores;
    } catch (error) {
      console.error('Get all stores error:', error);
      throw error;
    }
  },

  // Get products for a specific store
  getStoreProducts: async (storeId) => {
    try {
      const { products } = await this.getAllStoresAndProducts();
      return products.filter(product => product.storeId === storeId);
    } catch (error) {
      console.error('Get store products error:', error);
      throw error;
    }
  },

  // Check if user is owner of store
  isOwner: async (storeID, userID) => {
    try {
      const response = await api.get(`/stores/${storeID}/owners/${userID}/check`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        return false;
      }
    } catch (error) {
      return false;
    }
  },

  // Check if user is manager of store
  isManager: async (storeID, userID) => {
    try {
      const response = await api.get(`/stores/${storeID}/managers/${userID}/check`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        return false;
      }
    } catch (error) {
      return false;
    }
  },

  // Check if user is founder of store
  isFounder: async (storeID, userID) => {
    try {
      const response = await api.get(`/stores/${storeID}/founders/${userID}/check`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        return false;
      }
    } catch (error) {
      return false;
    }
  },

  // Close store
  closeStore: async (storeID, userName) => {
    try {
      const response = await api.post(`/stores/${storeID}/close?userName=${userName}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        throw new Error('Failed to close store');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to close store');
      }
      throw new Error('Failed to close store');
    }
  },

  // Open store
  openStore: async (storeID, userName) => {
    try {
      const response = await api.post(`/stores/${storeID}/open?userName=${userName}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        throw new Error('Failed to open store');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to open store');
      }
      throw new Error('Failed to open store');
    }
  },

  // Edit listing price
  editListingPrice: async (userName, storeID, listingId, newPrice) => {
    try {
      const response = await api.put('/stores/listings/price', {
        userName,
        storeID,
        listingId,
        newPrice
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Price updated successfully'
        };
      } else {
        throw new Error('Failed to update price');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to update price');
      }
      throw new Error('Failed to update price');
    }
  },

  // Edit listing product name
  editListingName: async (userName, storeID, listingId, newValue) => {
    try {
      const response = await api.put('/stores/listings/name', {
        userName,
        storeID,
        listingId,
        newValue
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Product name updated successfully'
        };
      } else {
        throw new Error('Failed to update product name');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to update product name');
      }
      throw new Error('Failed to update product name');
    }
  },

  // Edit listing description
  editListingDescription: async (userName, storeID, listingId, newValue) => {
    try {
      const response = await api.put('/stores/listings/description', {
        userName,
        storeID,
        listingId,
        newValue
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Description updated successfully'
        };
      } else {
        throw new Error('Failed to update description');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to update description');
      }
      throw new Error('Failed to update description');
    }
  },

  // Edit listing quantity
  editListingQuantity: async (userName, storeID, listingId, newQuantity) => {
    try {
      const response = await api.put('/stores/listings/quantity', {
        userName,
        storeID,
        listingId,
        newQuantity
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Quantity updated successfully'
        };
      } else {
        throw new Error('Failed to update quantity');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to update quantity');
      }
      throw new Error('Failed to update quantity');
    }
  },

  // Edit listing category
  editListingCategory: async (userName, storeID, listingId, newValue) => {
    try {
      const response = await api.put('/stores/listings/category', {
        userName,
        storeID,
        listingId,
        newValue
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Category updated successfully'
        };
      } else {
        throw new Error('Failed to update category');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to update category');
      }
      throw new Error('Failed to update category');
    }
  },

  // Add permission to manager
  addPermissionToManager: async (managerID, appointerID, permissionID, storeID) => {
    try {
      const response = await api.post('/stores/managers/permissions/add', {
        managerID,
        appointerID,
        permissionID,
        storeID
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return {
          success: true,
          message: 'Permission added successfully'
        };
      } else {
        throw new Error('Failed to add permission');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to add permission');
      }
      throw new Error('Failed to add permission');
    }
  },

  // Get manager permissions
  getManagersPermissions: async (managerID, whoIsAsking, storeID) => {
    try {
      const response = await api.get(`/stores/${storeID}/managers/${managerID}/permissions?whoIsAsking=${whoIsAsking}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        throw new Error('Failed to get permissions');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to get permissions');
      }
      throw new Error('Failed to get permissions');
    }
  },

  // Get current user's permissions and role in store
  getCurrentUserPermissions: async (storeID, userID) => {
    try {
      const response = await api.get(`/stores/${storeID}/user/${userID}/permissions`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        throw new Error('Failed to get user permissions');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to get user permissions');
      }
      throw new Error('Failed to get user permissions');
    }
  },

  // Get all store users (owners and managers)
  getStoreUsers: async (storeID, requesterId) => {
    try {
      const response = await api.get(`/stores/${storeID}/users?requesterId=${requesterId}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return apiResponse.data;
      } else {
        throw new Error('Failed to get store users');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to get store users');
      }
      throw new Error('Failed to get store users');
    }
  },

  // Add additional store owner
  addAdditionalStoreOwner: async (appointerID, newOwnerID, storeID) => {
    try {
      const response = await api.post('/stores/owners/add', {
        appointerID,
        newOwnerID,
        storeID
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return { success: true, message: 'Owner added successfully' };
      } else {
        throw new Error('Failed to add owner');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to add owner');
      }
      throw new Error('Failed to add owner');
    }
  },

  // Add new manager
  addNewManager: async (appointerID, newManagerName, storeID) => {
    try {
      const response = await api.post('/stores/managers/add', {
        appointerID,
        newManagerName,
        storeID
      });
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return { success: true, message: 'Manager added successfully' };
      } else {
        throw new Error('Failed to add manager');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to add manager');
      }
      throw new Error('Failed to add manager');
    }
  },

  // Remove owner
  removeOwner: async (requesterId, toRemove, storeID) => {
    try {
      const response = await api.delete(`/stores/${storeID}/owners/${toRemove}?requesterId=${requesterId}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return { 
          success: true, 
          message: 'Owner removed successfully',
          removedUsers: apiResponse.data 
        };
      } else {
        throw new Error('Failed to remove owner');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to remove owner');
      }
      throw new Error('Failed to remove owner');
    }
  },

  // Remove manager
  removeManager: async (appointerID, managerID, storeID) => {
    try {
      const response = await api.delete(`/stores/${storeID}/managers/${managerID}?appointerID=${appointerID}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return { success: true, message: 'Manager removed successfully' };
      } else {
        throw new Error('Failed to remove manager');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to remove manager');
      }
      throw new Error('Failed to remove manager');
    }
  },

  // Remove permission from manager
  removePermissionFromManager: async (managerID, permissionID, appointerID, storeID) => {
    try {
      const response = await api.delete(`/stores/${storeID}/managers/${managerID}/permissions/${permissionID}?appointerID=${appointerID}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return { success: true, message: 'Permission removed successfully' };
      } else {
        throw new Error('Failed to remove permission');
      }
    } catch (error) {
      if (error.response?.data?.error) {
        throw new Error('Failed to remove permission');
      }
      throw new Error('Failed to remove permission');
    }
  },

  // Validate if user exists in the system
  validateUserExists: async (userId) => {
    try {
      const response = await api.get(`/users/validate/${encodeURIComponent(userId)}`);
      const apiResponse = response.data;
      
      if (apiResponse.success) {
        return { exists: apiResponse.data.exists };
      } else {
        return { exists: false };
      }
    } catch (error) {
      console.error('Error validating user existence:', error);
      return { exists: false };
    }
  }
}; 